### PR TITLE
Bind target type, waiting for data consumer, fix storage declare options

### PR DESCRIPTION
**Pull request by dshvedchenko**

### CODE REVIEW

**Comment by dshvedchenko** [comment]
currently Waiting Consumer limited by number of messages, maybe it is need to add timeout field also

**Comment by dshvedchenko** [comment]
@FabioBatSilva 
Hello, is there any chances to review it next week ?
have a good weekend

**Comment by FabioBatSilva** [comment]
@dshvedchenko Sorry i didn't review it yet,

I'll try to check it out during the weekend or early next week..
Have a good weekend you too.

**Comment by FabioBatSilva** [comment]
I like the UI changes you did, 
But there are a lot of things do configure now, its getting a little confuse.
I was wondering if we should create a new tab for bindings, know do you think ?

File: src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java [hunk 1]
```diff
@@ -166,23 +170,59 @@ private void produce(Object[] r) throws IOException, KettleStepException
...
+			logDebug(consumerTag + " :SHUTDOWN: " + sig.getMessage() );
+		  }
+		};
+		channel.basicConsume(data.target, false, consumer);
+	}
         do {
-            response = channel.basicGet(data.target, false);
+	    if (data.isWaitingConsumer ) {
+		    QueueingConsumer.Delivery delivery = null;
+		    while (!isStopped() && delivery == null && !isStepTimedoutAsConsumer() ) {  delivery = consumer.nextDelivery(10000); } ;
```
Comment (FabioBatSilva) [inline src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java:193]: Why not this instead ?

``` java
try {
    consumer.nextDelivery(data.waitTimeout);
    // ...
} catch (InterruptedException e) {
    // ...
}
```

**Comment by FabioBatSilva** [comment]
Also I'm not sure how useful is the "Wait for" option,
As we discussed before something similar is possible using a "Wait for" step :

```
"Declare Queue/Bind Step" ->  "Wait for X Sec Step" -> "Queue Consume Step"
```

Cal you explain what is your use case for this ?

File: src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java (hunk 1 above)
Comment (dshvedchenko) [inline src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java:193]: data.waitTimeout  - is total amount of time for Waiting Consumer step can consume and wait for data from queue. consumer.nextDelivery(timeout) is time for waiting for one message.

**Comment by dshvedchenko** [comment]
I think it is more natural way for consume data from queue, probably costs less for CPU.
my use case, is to give more freedom for customers of plugin, to expose more features for RMQ client.
Specially marked them with red color to point they require more attention when used

File: src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java (hunk 1 above)
Comment (dshvedchenko) [inline src/main/java/com/instaclick/pentaho/plugin/amqp/AMQPPlugin.java:193]: | I was wondering if we should create a new tab for bindings, know do you think ?
yes, we should just don't have time to move them to separate tab.

**Comment by dshvedchenko** [comment]
so, current plans for pushing this changes into stable 2.0
- add tabs for Connections, Declare, Experimantal
- introduce option Nack ( re-queue for consumer mode ) , in order to give ability to check transformations without drain queue

**Comment by FabioBatSilva** [comment]
@dshvedchenko Cool, 

let me know when you are done and I'll review/merge it..

**Comment by dshvedchenko** [comment]
@FabioBatSilva 
Please, review latest changes

**Comment by dshvedchenko** [comment]
i've done with changes. now i`m waiting for comments

**Comment by FabioBatSilva** [comment]
@dshvedchenko Thanks

**Comment by dshvedchenko** [comment]
@FabioBatSilva 
Hello,
I've seen one commit ahead pull request. I have one concern.

removing 
            QueueingConsumer.Delivery delivery = null;
            while (!isStopped() && delivery == null && !isStepTimedoutAsConsumer() ) {  delivery = consumer.nextDelivery(10000); } ;

completely changed logic of WatingConsumer. Now (if queue is empty) it will wait for one message . My implementation was intended to limit lifetime of whole step. 
with current implementation this timeout is for time between two messages. if queue will be empty more than timeout, step finished. I think it worth to be mentioned in Readme

**Comment by FabioBatSilva** [comment]
Hi @dshvedchenko,

That is actually the expected behaviour of a waiting consumer,
But is certainly worth documenting the behaviour in the docs anyways.

**Comment by dbron0000** [comment]
Hi All,

Compiled the most recent snapshot on Windows (7 64bit) and Linux (ubuntu)
AMD64.

Configured a consumer on windows and bound to a queue using the wait for
data option. I'm finding that the first event consumed doesn't make it to
the next step (although subsequent events are logged (2nd step is write to
log).

Anyone seeing anything similar?

DB

**Comment by dshvedchenko** [comment]
Hi @dbron0000 

I will check today, will reply with info in 2 -4 hours

**Comment by dshvedchenko** [comment]
@dbron0000 
checked twice variant:
1. publish to queue
2. publish to exchange bound to queue

every time message proceed to second step immediately without loosing.

Please, attach gist with console output of run in rowlevel logging mode

**Comment by dbron0000** [comment]
Will do.

Retested with Linux client and not seeing the same behavior. May just be a
Windows issue on that workstation.

On Tue, Apr 7, 2015 at 11:24 AM, Denis Shvedchenko <notifications@github.com

**Pull request merged**